server/polar/order/service.py

```diff
@@ -1242,5 +1244,46 @@ async def _on_order_paid(self, session: AsyncSession, order: Order) -> None:
                 subscription_id=order.subscription_id,
             )
 
+    async def handle_payment_failure(
+        self, session: AsyncSession, order: Order
+    ) -> Order:
+        """Handle payment failure for an order, initiating dunning if necessary."""
+        if order.subscription is None:
+            return order
+
+        if order.subscription.stripe_subscription_id is not None:
+            # If the subscription is managed by Stripe, we don't handle dunning. Stripe will handle it.
+            return order
+
+        if order.next_payment_attempt_at is None:
+            return await self._handle_first_dunning_attempt(session, order)
+
+        return await self._handle_consecutive_dunning_attempts(session, order)
+
+    async def _handle_first_dunning_attempt(
+        self, session: AsyncSession, order: Order
+    ) -> Order:
+        """Handle the first dunning attempt for an order, setting the next payment
+        attempt date and marking the subscription as past due.
+        """
+
+        first_retry_date = utc_now() + settings.DUNNING_FIRST_RETRY_DELAY
```

Comment:
Code nitpick, but that may be nice to handle all the delays in a single list rather than having the first as a separate one.

---

I'm wondering then if we could factorize `_handle_first_dunning_attempt` and `_handle_consecutive_dunning_attempts` into a single method, where the first attempt would be a general case. Pseudo-code:

```
attempts = get_attempts(order)
if attempts >= len(settings.DUNNING_RETRY_INTERVALS):
    # Definitely failed
else:
    delay = settings.DUNNING_RETRY_INTERVALS[attempts - 1]
    next_attempt_at = (order.next_payment_attempt_at or utc_now()) + delay
```

But we might see that when we have the implementation of `_handle_consecutive_dunning_attempts` :)

Reply:
I was thinking about that, at first I wanted to keep it separate because it made my life easier to read but I may merge them in the next PR